### PR TITLE
Add kvm_stat case

### PR DESCRIPTION
**Pull request by yiqianwei**
ID: 1959718

Signed-off-by: Yiqian Wei <yiwei@redhat.com>

### CODE REVIEW

**Comment by yiqianwei** [comment]
(1/1) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.kvm_stat.q35: PASS (67.61 s)

**Comment by yiqianwei** [comment]
@yanan-fu @fbq815 Could you help to review this patch? Thanks

**Comment by fbq815** [comment]
LGTM and test result on s390x:
JOB ID     : 205900cc4cee93ebbd594a606fa0f9e621fdbf6c
JOB LOG    : /home/bfu/kar/workspace/job-results/job-2021-05-14T05.37-205900c/job.log
 (1/1) Host_RHEL.m8.u5.product_av.nographic.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.s390x.io-github-autotest-qemu.kvm_stat.s390-virtio: PASS (62.51 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/bfu/kar/workspace/job-results/job-2021-05-14T05.37-205900c/results.html
JOB TIME   : 63.06 s

File: qemu/tests/kvm_stat.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    """
+    interval = float(params.get("interval_time", "10"))
+
+    def kvm_stat_install_package():
+        kvm_stat_install_cmd = params.get("kvm_stat_install_cmd")
+        s = process.system(kvm_stat_install_cmd, timeout=3600, shell=True)
+        if s != 0:
+            test.error("Fail to install kvm_stat")
+        else:
+            logging.info("Install kvm_stat successfully.")
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:32]: Suggest:
```
from virttest import utils_package
...

utils_package.package_install(****)
```

File: qemu/tests/kvm_stat.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            result = line.strip().decode()
+            if 'kvm_stat.csv.5.gz' in result:
+                logging.info("Test complete")
+
+    kvm_stat_support_check_cmd = params.get("kvm_stat_support_check_cmd")
+    error_context.context("Check kvm_stat installed in host", logging.info)
+    try:
+        process.system(kvm_stat_support_check_cmd, timeout=interval, shell=True)
+    except Exception:
+        kvm_stat_install_package()
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:68]:
```
if not utils_package.package_install("kernel-tools"):
    test.cancel(*****)
```

Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: @yanan-fu ,According to your suggestion, I have changed it. Could you help to review this patch again ? thanks

(1/1) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.kvm_stat.q35: PASS (78.53 s)

**Comment by jingzhao84** [comment]
@yanan-fu  could you help to review it again, thanks

File: qemu/tests/kvm_stat.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        for line in out:
+            result = line.strip().decode()
+            if 'kvm_stat.csv.5.gz' in result:
+                logging.info("Test complete")
+
+    logging.info("Install kernel-tools package in host")
+    pkg_name = params.get("pkg_name")
+    utils_package.package_install(pkg_name)
+    if not utils_package.package_install("kernel-tools"):
+        test.cancel("Install dependency packages failed")
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:60]: Update L57~L59:
```
pkgs = params.objects("pkg_name")
if not utils_package.package_install(pkgs):
    test.cancel(****)
```

Comment (yiqianwei) [reply qemu/tests/kvm_stat.py:59]: HI Yanan,
I've updated it, Please review this patch again.thanks a lot.

**Comment by fbq815** [comment]
@yiqianwei @yanan-fu this patch has been delay for a really long time, could we please review this again and merge or close it?

File: qemu/tests/kvm_stat.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                logging.info("Test complete")
+
+    logging.info("Install kernel-tools package in host")
+    pkgs = params.objects("pkg_name")
+    if not utils_package.package_install(pkgs):
+        test.cancel("Install dependency packages failed")
+    logging.info("Adjust the parameter '-s' is 0.2 in the kvm_stat.service")
+    kvm_stat_service_path = params.get("kvm_stat_service_path")
+    if os.path.exists(kvm_stat_service_path):
+        tweak_file(kvm_stat_service_path, "10", "0.2")
```
[the remark this answers is not in the document]
Comment (yanan-fu) [reply qemu/tests/kvm_stat.py]: @yiqianwei  Waiting for your info, thanks!

File: qemu/tests/kvm_stat.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    start_service = process.system(kvm_stat_start_cmd, timeout=interval, shell=True)
+    if start_service != 0:
+        test.error("Failed to start the kvm_stat.server")
+    else:
+        logging.info("Successfully to start the kvm_stat.service")
+    params["start_vm"] = 'yes'
+    env_process.preprocess_vm(test, params, env, params.get("main_vm"))
+    vm = env.get_vm(params["main_vm"])
+    logging.info("Start logrotate command")
+    tweak_file("/etc/logrotate.d/kvm_stat", "10M", "10K")
```
[the remark this answers is not in the document]
Comment (yanan-fu) [reply qemu/tests/kvm_stat.py]: @yiqianwei  And here,  what's the answer for this issue ? Thanks!

[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: @yanan-fu,Yes, 10M works always

File: qemu/tests/kvm_stat.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: @yanan-fu ,the default value always be 10 for all platform/produce,There is no way to get the default value.

File: qemu/tests/kvm_stat.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        out = check_file_cmd.stdout.readlines()
+        for line in out:
+            result = line.strip().decode()
+            if 'kvm_stat.csv.5.gz' in result:
+                logging.info("Test complete")
+
+    depends_pkgs = params.objects("depends_pkgs")
+    logging.info("Install packages: %s in host", depends_pkgs)
+    if not utils_package.package_install(depends_pkgs):
+        test.cancel("Install dependency packages failed")
```
[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: I update it, Please check the latest code. thanks

File: qemu/tests/kvm_stat.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        for line in out:
+            result = line.strip().decode()
+            if 'kvm_stat.csv.5.gz' in result:
+                logging.info("Test complete")
+
+    logging.info("Install kernel-tools package in host")
+    pkgs = params.objects("pkg_name")
+    if not utils_package.package_install(pkgs):
+        test.cancel("Install dependency packages failed")
+    logging.info("Adjust the parameter '-s' is 0.2 in the kvm_stat.service")
```
[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: I update it, please check the latest code, thanks

**Comment by jingzhao84** [comment]
@yanan-fu 
Hello,  yfu
Could you help to check it?
Seems this patch stay for a long time
Many thanks
JIng

File: qemu/tests/kvm_stat.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    :param env: Dictionary with test environment
+    """
+    interval = float(params.get("interval_time", "10"))
+
+    def tweak_file(filename, old_str, new_str):
+        # replace the strings in the file
+        file_data = ""
+        with open(filename, "r") as f:
+            for line in f:
+                if old_str in line:
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:32]: What's the expected result if no `old_str` in the file content ?  Currently, the new value will not be set and test will continue, i think it is not suitable.

File: qemu/tests/cfg/kvm_stat.cfg [hunk 9]
```diff
@@ -0,0 +1,11 @@
+- kvm_stat:
+    virt_test_type = qemu
+    type = kvm_stat
+    depends_pkgs = "kernel-tools"
+    kvm_stat_service_path = "/usr/lib/systemd/system/kvm_stat.service"
+    kvm_stat_interval = 0.2
+    kvm_stat_start_cmd = "systemctl start kvm_stat.service"
```
Comment (yanan-fu) [inline qemu/tests/cfg/kvm_stat.cfg:7]: Suggest using `restart` to ensure service be restarted after the parameter be changed

File: qemu/tests/kvm_stat.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    else:
+        logging.info("Successfully to start the kvm_stat.service")
+    params["start_vm"] = 'yes'
+    env_process.preprocess_vm(test, params, env, params.get("main_vm"))
+    vm = env.get_vm(params["main_vm"])
+    logging.info("Start logrotate command")
+    file_size = params.get("file_size")
+    tweak_file("/etc/logrotate.d/kvm_stat", "10M", file_size)
+    log_path = params.get("log_path")
+    utils_misc.wait_for(lambda: separate_file(log_path), 40)
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:80]: From manual case, there is one step to check log file number do not more than `rotate` in `/etc/logrotate.d/kvm_stat`.
If you are using `while num < 6:` in `separate_file`, it is not make sense.

File: qemu/tests/kvm_stat.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+                    line = line.replace(old_str, new_str, 1)
+                file_data += line
+        with open(filename, "w") as f:
+            f.write(file_data)
+
+    def separate_file(file_path):
+        num = 1
+        file_size = os.path.getsize(file_path)
+        while num < 6:
+            if file_size > 10240:
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:42]: `10240` is the new `file_size` that be setted in `/etc/logrotate.d/kvm_stat`, do not hard code it here.

File: qemu/tests/cfg/kvm_stat.cfg (hunk 9 above)
Comment (yiqianwei) [reply qemu/tests/cfg/kvm_stat.cfg]: ok, I updated it .thanks

**Comment by yiqianwei** [comment]
@yanan-fu , please review this patch, I have updated it. Thanks

File: qemu/tests/kvm_stat.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+    interval = float(params.get("interval_time", "10"))
+
+    def tweak_file(origin, target, filename):
+        with open(filename, "r") as f:
+            content = f.read()
+            content = re.sub(origin, target, content)
+            f.close()
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:28]: No need, it can be closed automatically

File: qemu/tests/kvm_stat.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        with open(filename, "w") as f:
+            f.write(content)
+            f.close()
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:31]: Same, no need

File: qemu/tests/kvm_stat.py (hunk 12 above)
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: ok, I already remove 28 line

File: qemu/tests/kvm_stat.py (hunk 13 above)
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: ok, I already remove 31 line

File: qemu/tests/cfg/kvm_stat.cfg [hunk 14]
```diff
@@ -0,0 +1,12 @@
+- kvm_stat:
+    virt_test_type = qemu
+    type = kvm_stat
+    depends_pkgs = "kernel-tools"
+    kvm_stat_service_path = "/usr/lib/systemd/system/kvm_stat.service"
+    kvm_stat_interval = "0.2"
+    kvm_stat_start_cmd = "systemctl restart kvm_stat.service"
+    start_vm = no
+    logrotate_config_file_path = "/etc/logrotate.d/kvm_stat"
+    rotate_size_limi  = "10K"
```
Comment (yanan-fu) [inline qemu/tests/cfg/kvm_stat.cfg:10]: typo

File: qemu/tests/kvm_stat.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+    def separate_file(file_path):
+        f = open(logrotate_config_file_path, "r")
+        content = f.read()
+        rotate_time = int(re.findall(r"rotate.(\d?)", content)[0]) + 1
+        f.close()
+        file_size = os.path.getsize(file_path)
+        num = 1
+        while num <= rotate_time:
+            if file_size > int(rotate_size_limit[0:-1]) * 1024:
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:39]: Suggest `from virttest.utils_numeric import normalize_data_size`

File: qemu/tests/kvm_stat.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+                process.system('logrotate -v /etc/logrotate.d/kvm_stat')
+            num += 1
+
+    def check_log():
+        check_log_cmd = params.get("check_log_cmd")
+        cmd = subprocess.Popen(check_log_cmd, shell=True, stdout=subprocess.PIPE)
+        results = cmd.stdout.readlines()
+        for line in results:
+            results = line.strip().decode()
+        re.findall(r"kvm_stat.csv.\d.gz", results)
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:49]: What's the expected result ?   I did not see any check point.

File: qemu/tests/cfg/kvm_stat.cfg (hunk 14 above)
Comment (yiqianwei) [reply qemu/tests/cfg/kvm_stat.cfg]: update :
 rotate_size_limit = "10K"

File: qemu/tests/kvm_stat.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+            f.close()
+
+    def separate_file(file_path):
+        f = open(logrotate_config_file_path, "r")
+        content = f.read()
+        rotate_time = int(re.findall(r"rotate.(\d?)", content)[0]) + 1
+        f.close()
+        file_size = os.path.getsize(file_path)
+        num = 1
+        while num <= rotate_time:
```
[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: ok , done

File: qemu/tests/kvm_stat.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+            content = re.sub(origin, target, content)
+            f.close()
+        with open(filename, "w") as f:
+            f.write(content)
+            f.close()
+
+    def separate_file(file_path):
+        f = open(logrotate_config_file_path, "r")
+        content = f.read()
+        rotate_time = int(re.findall(r"rotate.(\d?)", content)[0]) + 1
```
[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: remove 34 ~46 line, set "rotate_time = 5 " to config file

File: qemu/tests/kvm_stat.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+        f.close()
```
[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: remove it

File: qemu/tests/kvm_stat.py (hunk 15 above)
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: update:

if file_size > int(normalize_data_size(rotate_size_limit, order_magnitude="B", factor=1024)):

**Comment by yiqianwei** [comment]
@yanan-fu, I have updated it,  please review this patch. thanks

File: qemu/tests/kvm_stat.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: @yanan-fu, My understanding is wrong, can I change the following code?
    def separate_file(file_path):
        with open(logrotate_config_file_path, "r") as f:
            content = f.read()
            rotate_time = int(re.search('rotate\s+(\d+)', content)[1])
        file_size = os.path.getsize(file_path)
        num = rotate_time + 1
        while num > 0:
        ...

[the remark this answers is not in the document]
Comment (yanan-fu) [reply qemu/tests/kvm_stat.py]: `rotata_time` be used in both `separate_file` and `check_log`, need move it out.

[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: yes, Please check the latest code, lines 82 ~ 84

File: qemu/tests/kvm_stat.py [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+    test.log.info("Start logrotate command")
+    logrotate_config_file_path = params.get("logrotate_config_file_path")
+    rotate_size_limit = params.get("rotate_size_limit")
+    tweak_file(r"size\s\d+M", "size %s" % rotate_size_limit, logrotate_config_file_path)
+    log_file = params.get("log_file")
+    with open(logrotate_config_file_path, "r") as f1:
+        content1 = f1.read()
+        rotate_time = int(re.search(r"rotate\s+(\d+)", content1)[1])
+    utils_misc.wait_for(separate_file(log_file), 45)
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:85]: It is wrong for `wait_for` ... suggest:
```
    def separate_file(file_path, rotate_size_limit):
        rotate_size_limit = int(normalize_data_size(rotate_size_limit, order_magnitude="B",
                                                    factor=1024))
        num = rotate_time + 1
        while num > 0:
            if utils_misc.wait_for(lambda: os.path.getsize(file_path) > rotate_size_limit, 60, 5, 5):
                process.system('logrotate -v /etc/logrotate.d/kvm_stat')
            num -= 1
```

File: qemu/tests/kvm_stat.py [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+    depends_pkgs = params.objects("depends_pkgs")
+    test.log.info("Install packages: %s in host", depends_pkgs)
+    if not utils_package.package_install(depends_pkgs):
+        test.cancel("Install %s packages failed", depends_pkgs)
+
+    kvm_stat_service_path = params.get("kvm_stat_service_path")
+    kvm_stat_interval = params.get("kvm_stat_interval")
+    test.log.info("Adjust the parameter '-s' is %s in %s", kvm_stat_interval, kvm_stat_service_path)
+    if os.path.exists(kvm_stat_service_path):
+        tweak_file(r"-s\s\d+", "-s %s" % kvm_stat_interval, kvm_stat_service_path)
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:61]:
```
    if os.path.exists(kvm_stat_service_path):
        with open(kvm_stat_service_path, 'r') as fd:
            content = fd.read()
        initial_kvm_stat_interval = re.search('ExecStart=.*-s\s([\.\d]+)\s', content).group(1)
        tweak_file(r"-s %s" % initial_kvm_stat_interval, "-s %s" % kvm_stat_interval, kvm_stat_service_path)
    else:
        test.error("%s file does not exist", kvm_stat_service_path)
```

File: qemu/tests/kvm_stat.py [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+        test.log.info("Successfully to start the kvm_stat.service")
+
+    params["start_vm"] = 'yes'
+    env_process.preprocess_vm(test, params, env, params.get("main_vm"))
+    vm = env.get_vm(params["main_vm"])
+
+    test.log.info("Start logrotate command")
+    logrotate_config_file_path = params.get("logrotate_config_file_path")
+    rotate_size_limit = params.get("rotate_size_limit")
+    tweak_file(r"size\s\d+M", "size %s" % rotate_size_limit, logrotate_config_file_path)
```
Comment (yanan-fu) [inline qemu/tests/kvm_stat.py:80]:
```
    logrotate_config_file_path = params.get("logrotate_config_file_path")
    with open(logrotate_config_file_path, 'r') as fd:
        content = fd.read()
    initial_rotate_size = re.search('size\s(.*)\s', content).group(1)
    rotate_size_limit = params.get("rotate_size_limit")
    tweak_file(r"size %s" % initial_rotate_size, "size %s" % rotate_size_limit, logrotate_config_file_path)
```

File: qemu/tests/kvm_stat.py [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+    log_file = params.get("log_file")
+    with open(logrotate_config_file_path, "r") as f1:
+        content1 = f1.read()
+        rotate_time = int(re.search(r"rotate\s+(\d+)", content1)[1])
+    utils_misc.wait_for(separate_file(log_file), 45)
+
+    test.log.info("Check rotate log file")
+    check_log()
```
[the remark this answers is not in the document]
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py:96]: OK, please check lines 52~54 in the latest code

File: qemu/tests/kvm_stat.py (hunk 22 above)
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: done, thanks

File: qemu/tests/kvm_stat.py (hunk 21 above)
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: OK, I have updated it. thanks

File: qemu/tests/kvm_stat.py (hunk 20 above)
Comment (yiqianwei) [reply qemu/tests/kvm_stat.py]: OK, I have updated it, thanks

**Comment by yiqianwei** [comment]
@yanan-fu ,  I have updated it. Thanks

(1/1) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.1.0.x86_64.io-github-autotest-qemu.kvm_stat.q35: PASS (87.89 s)

**Comment by yiqianwei** [comment]
> Restore host env should be executed always, please use `try .. finally`

@yanan-fu, I have updated it, please check code

**Comment by yiqianwei** [comment]
> @yiqianwei Call for latest test result, thanks!
Auto test results:
(1/1) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.1.0.x86_64.io-github-autotest-qemu.kvm_stat.q35: PASS (83.12 s)

**Review by yanan-fu — APPROVED** [review]
Ack

**Pull request merged**